sdks/python/apache_beam/pipeline.py

```diff
@@ -157,6 +157,113 @@ def _root_transform(self):
+          inputs = transform_node.inputs
+          # We only support replacing single-input PTransforms.
+          assert len(inputs) == 1
+          transform_node.transform = replacement_transform
+          self.pipeline.transforms_stack.append(transform_node)
+
+          # Keeping the same label for the replaced node but recursively
+          # removing labels of child transforms since they will be replaced
+          # during the expand below.
+          self.pipeline._remove_labels_recursively(transform_node)
+
+          new_output = replacement_transform.expand(inputs[0])
+
+          # Recording updated outputs. This cannot be done in the same visor
+          # since if we dynamically update output type here, we'll run into
+          # errors when visiting child nodes.
+          output_map[transform_node.outputs[None]] = new_output
+
+          self.pipeline.transforms_stack.pop()
+
+      def enter_composite_transform(self, transform_node):
+        self._replace_if_needed(transform_node)
+
+      def visit_transform(self, transform_node):
+        self._replace_if_needed(transform_node)
+
+    self.visit(OutputVisitor(self, self.applied_labels))
+
+    # Adjusting inputs and outputs
+    class OutputUpdater(PipelineVisitor): # pylint: disable=used-before-assignment
+
+      def __init__(self, pipeline):
+        self.pipeline = pipeline
+
+      def visit_transform(self, transform_node):
+        if transform_node.outputs[None] in output_map:
+          transform_node.replace_output(
+              output_map[transform_node.outputs[None]])
+
+        replace_input = False
+        for input in transform_node.inputs:
+          if input in output_map:
+            replace_input = True
+            break
+
+        if replace_input:
+          transform_node.inputs = [
+              input if not input in output_map else output_map[input]
+              for input in transform_node.inputs]
+
+    self.visit(OutputUpdater(self))
+
+  def _check_replacement(self, override):
+    matcher = override.get_matcher()
+
+    class Visitor(PipelineVisitor):
+      def visit_transform(self, transform_node):
+        if matcher.match(transform_node):
+          raise RuntimeError('Transform node %r was not replaced as expected.',
+                             transform_node)
+
+    self.visit(Visitor())
+
+  def replace_all(self, replacements):
+    """ Dynamically replaces PTransforms in the currently populated hierarchy.
+
+     Currently this only works for replacements where input and output types
+     are exactly the same.
+     TODO: Update this to also work for transform overrides where input and
+     output types are different.
+
+    Args:
+      replacements a list of PTransformOverride objects.
+    """
+    for override in replacements:
+      assert isinstance(override, PTransformOverride)
+      self._replace(override)
+
+    # Checking if the transforms have been successfully replaced.
+    for override in replacements:
```

Comment:
Note about when this could happen (e.g. the ordering of replacements is important, right?)